From: Ben Hutchings <ben@decadent.org.uk>
Subject: [PATCH] Change configure script to use Kbuild for compile tests
Last-Update: 2010-05-23

Kbuild is the only proper way to build modules.  Calling the compiler
directly will tend to result in false negatives.

Index: alsa-driver-1.0.25/configure.in
===================================================================
--- alsa-driver-1.0.25.orig/configure.in	2012-02-08 20:22:24.146529654 +1100
+++ alsa-driver-1.0.25/configure.in	2012-02-08 20:25:24.334534269 +1100
@@ -1702,11 +1702,6 @@
 fi
 AC_SUBST(msmp)
 
-KERNEL_CHECK_CFLAGS="-Wall -fno-strict-aliasing $CFLAGS $KERNEL_INC $HACK_KERNEL_INC -Iinclude -nostdinc -iwithprefix include"
-if test "$NEW_KBUILD" = "y"; then
-  KERNEL_CHECK_CFLAGS="$KERNEL_CHECK_CFLAGS -DKBUILD_STR(s)=#s -DKBUILD_BASENAME=KBUILD_STR(dummy) -DKBUILD_MODNAME=KBUILD_BASENAME"
-fi
-
 dnl Check for video device support... (2.2 kernels excluded due to API incompatibility)
 if test $kvernum -gt 202000; then
 CHECK_KERNEL_CONFIG(CONFIG_VIDEO_DEV, [Video device support in kernel])
@@ -1783,20 +1778,32 @@
 fi
 AC_SUBST(CONFIG_PNP)
 
+dnl Replacement for AC_TRY_COMPILE using kbuild
+AC_DEFUN([KBUILD_TRY_COMPILE], [
+dir="$(mktemp -d)"
+if test -z "$dir"; then
+  echo >&2 "failed to create temporary directory"
+  exit 1
+fi
+echo >"$dir"/test.c '[$1]
+void not_main(void)
+{
+	[$2];
+}'
+echo >"$dir"/Makefile 'obj-y += test.o
+EXTRA_CFLAGS := [$5]'
+if make -C "$CONFIG_SND_KERNELDIR" SUBDIRS="$dir" >&5 2>&1; then
+  [$3]
+else
+  [$4]
+fi
+rm -rf "$dir"
+])
+
 dnl Check for strlcpy...
 AC_MSG_CHECKING(for strlcpy)
 strlcpy="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kernel.h>
 #include <linux/string.h>
 ],[
@@ -1805,10 +1812,8 @@
 ],
     AC_MSG_RESULT(yes);strlcpy="1",
     AC_MSG_RESULT(no);strlcpy="0",
-    AC_MSG_RESULT(unknown);strlcpy="0"
+    -Werror
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_STRLCPY=$strlcpy
 dnl AC_SUBST(CONFIG_HAVE_STRLCPY)
 if test "$CONFIG_HAVE_STRLCPY" = "1"; then
@@ -1818,17 +1823,7 @@
 dnl Check for snprintf...
 AC_MSG_CHECKING(for snprintf)
 snprintf="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kernel.h>
 ],[
   char buf[128];
@@ -1836,10 +1831,8 @@
 ],
     AC_MSG_RESULT(yes);snprintf="1",
     AC_MSG_RESULT(no);snprintf="0",
-    AC_MSG_RESULT(unknown);snprintf="0"
+    -Werror
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_SNPRINTF=$snprintf
 dnl AC_SUBST(CONFIG_HAVE_SNPRINTF)
 if test "$CONFIG_HAVE_SNPRINTF" = "1"; then
@@ -1849,17 +1842,7 @@
 dnl Check for vsnprintf...
 AC_MSG_CHECKING(for vsnprintf)
 vsnprintf="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kernel.h>
 void f(const char *fmt, ...)
 {
@@ -1874,10 +1857,8 @@
 ],
     AC_MSG_RESULT(yes);vsnprintf="1",
     AC_MSG_RESULT(no);vsnprintf="0",
-    AC_MSG_RESULT(unknown);vsnprintf="0"
+    -Werror
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_VSNPRINTF=$vsnprintf
 dnl AC_SUBST(CONFIG_HAVE_VSNPRINTF)
 if test "$CONFIG_HAVE_VSNPRINTF" = "1"; then
@@ -1887,17 +1868,7 @@
 dnl Check for scnprintf...
 AC_MSG_CHECKING(for scnprintf)
 scnprintf="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kernel.h>
 ],[
   char buf[128];
@@ -1905,10 +1876,8 @@
 ],
     AC_MSG_RESULT(yes);scnprintf="1",
     AC_MSG_RESULT(no);scnprintf="0",
-    AC_MSG_RESULT(unknown);scnprintf="0"
+    -Werror
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_SCNPRINTF=$scnprintf
 dnl AC_SUBST(CONFIG_HAVE_SCNPRINTF)
 if test "$CONFIG_HAVE_SCNPRINTF" = "1"; then
@@ -1918,17 +1887,7 @@
 dnl Check for sscanf...
 AC_MSG_CHECKING(for sscanf)
 sscanf="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kernel.h>
 ],[
   unsigned int i;
@@ -1936,10 +1895,8 @@
 ],
     AC_MSG_RESULT(yes);sscanf="1",
     AC_MSG_RESULT(no);sscanf="0",
-    AC_MSG_RESULT(unknown);sscanf="0"
+    -Werror
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_SSCANF=$sscanf
 dnl AC_SUBST(CONFIG_HAVE_SSCANF)
 if test "$CONFIG_HAVE_SSCANF" = "1"; then
@@ -1949,17 +1906,7 @@
 dnl Check for vmalloc_to_page...
 AC_MSG_CHECKING(for vmalloc_to_page)
 vmalloc_to_page="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/time.h>
 #include <linux/sched.h>
 #include <linux/interrupt.h>
@@ -1970,10 +1917,7 @@
 ],
     AC_MSG_RESULT(yes);vmalloc_to_page="1",
     AC_MSG_RESULT(no);vmalloc_to_page="0",
-    AC_MSG_RESULT(unknown);vmalloc_to_page="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_VMALLOC_TO_PAGE=$vmalloc_to_page
 dnl AC_SUBST(CONFIG_HAVE_VMALLOC_TO_PAGE)
 if test "$CONFIG_HAVE_VMALLOC_TO_PAGE" = "1"; then
@@ -2018,27 +1962,14 @@
 dnl Check for old kmod...
 AC_MSG_CHECKING(for old kmod)
 old_kmod="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/kmod.h>
 ],[
 	request_module("abcd", "def", "ixj");
 ],
     AC_MSG_RESULT(no);old_kmod="0",
     AC_MSG_RESULT(yes);old_kmod="1",
-    AC_MSG_RESULT(unknown);old_kmod="1"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_OLD_REQUEST_MODULE=$old_kmod
 dnl AC_SUBST(CONFIG_HAVE_OLD_REQUEST_MODULE)
 if test "$CONFIG_HAVE_OLD_REQUEST_MODULE" = "1"; then
@@ -2048,17 +1979,7 @@
 dnl Check for PDE...
 AC_MSG_CHECKING(for PDE)
 pde_defined="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/fs.h>
 #include <linux/proc_fs.h>
 ],[
@@ -2067,10 +1988,7 @@
 ],
     AC_MSG_RESULT(yes);pde_defined="1",
     AC_MSG_RESULT(no);pde_defined="0",
-    AC_MSG_RESULT(unknown);pde_defined="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PDE=$pde_defined
 dnl AC_SUBST(CONFIG_HAVE_PDE)
 if test "$CONFIG_HAVE_PDE" = "1"; then
@@ -2080,17 +1998,7 @@
 dnl Check for pci_consistent_set_dma_mask()
 AC_MSG_CHECKING(for pci_set_consistent_dma_mask)
 pci_consistent_defined="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/pci.h>
 ],[
 	int (*func)();
@@ -2098,10 +2006,7 @@
 ],
     AC_MSG_RESULT(yes);pci_consistent_defined="1",
     AC_MSG_RESULT(no);pci_consistent_defined="0",
-    AC_MSG_RESULT(unknown);pci_consistent_defined="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PCI_CONSISTENT_DMA_MASK=$pci_consistent_defined
 if test "$CONFIG_HAVE_PCI_CONSISTENT_DMA_MASK" = "1"; then
   AC_DEFINE(CONFIG_HAVE_PCI_CONSISTENT_DMA_MASK)
@@ -2110,17 +2015,7 @@
 dnl Check for pci_set_dev_present()
 AC_MSG_CHECKING(for pci_dev_present)
 pci_consistent_defined="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/pci.h>
 ],[
 	int (*func)();
@@ -2128,10 +2023,7 @@
 ],
     AC_MSG_RESULT(yes);pci_dev_present_defined="1",
     AC_MSG_RESULT(no);pci_dev_present_defined="0",
-    AC_MSG_RESULT(unknown);pci_dev_present_defined="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PCI_DEV_PRESENT=$pci_dev_present_defined
 if test "$CONFIG_HAVE_PCI_DEV_PRESENT" = "1"; then
   AC_DEFINE(CONFIG_HAVE_PCI_DEV_PRESENT)
@@ -2140,17 +2032,7 @@
 dnl Check for msleep
 AC_MSG_CHECKING(for msleep)
 msleep="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/delay.h>
 ],[
   void (*func)(unsigned int);
@@ -2158,10 +2040,7 @@
 ],
     AC_MSG_RESULT(yes);msleep="1",
     AC_MSG_RESULT(no);msleep="0",
-    AC_MSG_RESULT(unknown);msleep="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_MSLEEP=$msleep
 dnl AC_SUBST(CONFIG_HAVE_MSLEEP)
 if test "$CONFIG_HAVE_MSLEEP" = "1"; then
@@ -2172,17 +2051,7 @@
 if test "$CONFIG_HAVE_MSLEEP" = "1"; then
 AC_MSG_CHECKING(for msleep_interrupt)
 msleep_interrupt="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/delay.h>
 ],[
   unsigned long (*func)(unsigned int);
@@ -2190,10 +2059,7 @@
 ],
     AC_MSG_RESULT(yes);msleep_interruptible="1",
     AC_MSG_RESULT(no);msleep_interruptible="0",
-    AC_MSG_RESULT(unknown);msleep_interruptible="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_MSLEEP_INTERRUPTIBLE=$msleep_interruptible
 dnl AC_SUBST(CONFIG_HAVE_MSLEEP_INTERRUPTIBLE)
 if test "$CONFIG_HAVE_MSLEEP_INTERRUPTIBLE" = "1"; then
@@ -2204,17 +2070,7 @@
 dnl Check for msecs_to_jiffies
 AC_MSG_CHECKING(for msecs_to_jiffies)
 msecs_to_jiffies="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Iinclude"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/jiffies.h>
 #include <linux/time.h>
 ],[
@@ -2223,10 +2079,8 @@
 ],
     AC_MSG_RESULT(yes);msecs_to_jiffies="1",
     AC_MSG_RESULT(no);msecs_to_jiffies="0",
-    AC_MSG_RESULT(unknown);msecs_to_jiffies="0"
+    -Iinclude
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_MSECS_TO_JIFFIES=$msecs_to_jiffies
 dnl AC_SUBST(CONFIG_HAVE_MSECS_TO_JIFFIES)
 if test "$CONFIG_HAVE_MSECS_TO_JIFFIES" = "1"; then
@@ -2236,17 +2090,7 @@
 dnl Check for tty->count is the atomic type
 AC_MSG_CHECKING(for tty->count is the atomic type)
 tty_count_atomic="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/tty.h>
 ],[
 	struct tty_struct tty;
@@ -2254,10 +2098,7 @@
 ],
     AC_MSG_RESULT(yes);tty_count_atomic="1",
     AC_MSG_RESULT(no);tty_count_atomic="0",
-    AC_MSG_RESULT(unknown);tty_count_atomic="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_TTY_COUNT_ATOMIC=$tty_count_atomic
 if test "$CONFIG_HAVE_TTY_COUNT_ATOMIC" = "1"; then
   AC_DEFINE(CONFIG_HAVE_TTY_COUNT_ATOMIC)
@@ -2285,10 +2126,8 @@
 ],
     AC_MSG_RESULT(yes);video_drvdata="1",
     AC_MSG_RESULT(no);video_drvdata="0",
-    AC_MSG_RESULT(unknown);video_drvdata="0"
+    -Iinclude
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_VIDEO_DRVDATA=$video_drvdata
 dnl AC_SUBST(CONFIG_HAVE_VIDEO_DRVDATA)
 if test "$CONFIG_HAVE_VIDEO_DRVDATA" = "1"; then
@@ -2322,17 +2161,7 @@
 dnl Check for old io_remap_page_range()
 AC_MSG_CHECKING(for io_remap_pfn_range)
 io_remap_pfn_range="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror-implicit-function-declaration"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/mm.h>
 ],[
   static pgprot_t p;
@@ -2340,10 +2169,8 @@
 ],
     AC_MSG_RESULT(yes);io_remap_pfn_range="1",
     AC_MSG_RESULT(no);io_remap_pfn_range="0",
-    AC_MSG_RESULT(unknown);io_remap_pfn_range="0"
+    -Werror-implicit-function-declaration
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 if test "$io_remap_pfn_range" = "1"; then
   AC_DEFINE(CONFIG_HAVE_IO_REMAP_PFN_RANGE)
 fi
@@ -2351,17 +2178,7 @@
 if test "$io_remap_pfn_range" != "1"; then
 AC_MSG_CHECKING(for new io_remap_page_range)
 new_io_remap="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS -Werror-implicit-function-declaration"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/mm.h>
 ],[
   static pgprot_t p;
@@ -2369,10 +2186,8 @@
 ],
     AC_MSG_RESULT(yes);new_io_remap="1",
     AC_MSG_RESULT(no);new_io_remap="0",
-    AC_MSG_RESULT(unknown);new_io_remap="0"
+    -Werror-implicit-function-declaration
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 if test "$new_io_remap" != "1"; then
   AC_DEFINE(CONFIG_OLD_IO_REMAP_PAGE_RANGE)
 fi
@@ -2381,17 +2196,7 @@
 dnl Check for kcalloc()
 AC_MSG_CHECKING(for kcalloc)
 kcalloc="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/slab.h>
 ],[
   void * (*func)(size_t, size_t, int);
@@ -2399,10 +2204,7 @@
 ],
     AC_MSG_RESULT(yes);kcalloc="1",
     AC_MSG_RESULT(no);kcalloc="0",
-    AC_MSG_RESULT(unknown);kcalloc="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_KCALLOC=$kcalloc
 dnl AC_SUBST(CONFIG_HAVE_KCALLOC)
 if test "$CONFIG_HAVE_KCALLOC" = "1"; then
@@ -2412,17 +2214,7 @@
 dnl Check for kstrdup()
 AC_MSG_CHECKING(for kstrdup)
 kstrdup="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/string.h>
 ],[
   char * (*func)(const char *, int);
@@ -2430,10 +2222,7 @@
 ],
     AC_MSG_RESULT(yes);kstrdup="1",
     AC_MSG_RESULT(no);kstrdup="0",
-    AC_MSG_RESULT(unknown);kstrdup="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_KSTRDUP=$kstrdup
 dnl AC_SUBST(CONFIG_HAVE_KSTRDUP)
 if test "$CONFIG_HAVE_KSTRDUP" = "1"; then
@@ -2443,13 +2232,7 @@
 dnl Check for kstrdup()
 AC_MSG_CHECKING(for kstrndup)
 kstrdup="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#include <linux/config.h>
+KBUILD_TRY_COMPILE([
 #include <linux/string.h>
 ],[
   char * (*func)(const char *, int);
@@ -2457,10 +2240,7 @@
 ],
     AC_MSG_RESULT(yes);kstrndup="1",
     AC_MSG_RESULT(no);kstrndup="0",
-    AC_MSG_RESULT(unknown);kstrndup="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_KSTRNDUP=$kstrndup
 dnl AC_SUBST(CONFIG_HAVE_KSTNRDUP)
 if test "$CONFIG_HAVE_KSTRNDUP" = "1"; then
@@ -2470,17 +2250,7 @@
 dnl Check for kzalloc()
 AC_MSG_CHECKING(for kzalloc)
 kzalloc="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/slab.h>
 ],[
   void * (*func)(size_t, int);
@@ -2488,10 +2258,7 @@
 ],
     AC_MSG_RESULT(yes);kzalloc="1",
     AC_MSG_RESULT(no);kzalloc="0",
-    AC_MSG_RESULT(unknown);kzalloc="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_KZALLOC=$kzalloc
 dnl AC_SUBST(CONFIG_HAVE_KZALLOC)
 if test "$CONFIG_HAVE_KZALLOC" = "1"; then
@@ -2502,27 +2269,14 @@
 if test $kvernum -ge 206000; then
 AC_MSG_CHECKING(for create_workqueue with flags)
 workqueue2="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/workqueue.h>
 ],[
     create_workqueue("abcd", 0);
 ],
     AC_MSG_RESULT(yes);workqueue2="1",
     AC_MSG_RESULT(no);workqueue2="0",
-    AC_MSG_RESULT(unknown);workqueue2="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_CREATE_WORKQUEUE_FLAGS=$workqueue2
 dnl AC_SUBST(CONFIG_CREATE_WORKQUEUE_FLAGS)
 if test "$CONFIG_CREATE_WORKQUEUE_FLAGS" = "1"; then
@@ -2533,17 +2287,7 @@
 dnl Check for pci_ioremap_bar
 AC_MSG_CHECKING(for pci_ioremap_bar)
 pci_ioremap_bar="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/pci.h>
 ],[
   void (*func)(struct pci_dev *, int);
@@ -2551,30 +2295,17 @@
 ],
     AC_MSG_RESULT(yes);pci_ioremap_bar="1",
     AC_MSG_RESULT(no);pci_ioremap_bar="0",
-    AC_MSG_RESULT(unknown);pci_ioremap_bar="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PCI_IOREMAP_BAR=$pci_ioremap_bar
 dnl AC_SUBST(CONFIG_HAVE_PCI_IOREMAP_BAR)
 if test "$CONFIG_HAVE_PCI_IOREMAP_BAR" = "1"; then
   AC_DEFINE(CONFIG_HAVE_PCI_IOREMAP_BAR)
 fi
-
-dnl Check for saved_config_space in pci_dev
-AC_MSG_CHECKING(for saved_config_space in pci_dev)
-pci_saved_config="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+
+dnl Check for saved_config_space in pci_dev
+AC_MSG_CHECKING(for saved_config_space in pci_dev)
+pci_saved_config="0"
+KBUILD_TRY_COMPILE([
 #include <linux/pci.h>
 ],[
   struct pci_dev dev;
@@ -2582,10 +2313,7 @@
 ],
     AC_MSG_RESULT(yes);pci_saved_config="1",
     AC_MSG_RESULT(no);pci_saved_config="0",
-    AC_MSG_RESULT(unknown);pci_saved_config="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PCI_SAVED_CONFIG=$pci_saved_config
 dnl AC_SUBST(CONFIG_HAVE_PCI_SAVED_CONFIG)
 if test "$CONFIG_HAVE_PCI_SAVED_CONFIG" = "1"; then
@@ -2596,17 +2324,7 @@
 dnl Check for new pci_save_state()
 AC_MSG_CHECKING(for new pci_save_state)
 new_pci_save_state="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/pci.h>
 ],[
   struct pci_dev *pci;
@@ -2614,10 +2332,7 @@
 ],
     AC_MSG_RESULT(yes);new_pci_save_state="1",
     AC_MSG_RESULT(no);new_pci_save_state="0",
-    AC_MSG_RESULT(unknown);new_pci_save_state="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 if test "$new_pci_save_state" = "1"; then
   AC_DEFINE(CONFIG_HAVE_NEW_PCI_SAVE_STATE)
 fi
@@ -2626,17 +2341,7 @@
 dnl Check for register_sound_special_device
 AC_MSG_CHECKING(for register_sound_special_device)
 msleep="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/sound.h>
 ],[
   void (*func)(struct file_operations *, int, struct device *);
@@ -2644,10 +2349,7 @@
 ],
     AC_MSG_RESULT(yes);reg_sp="1",
     AC_MSG_RESULT(no);reg_sp="0",
-    AC_MSG_RESULT(unknown);reg_sp="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_REGISTER_SOUND_SPECIAL_DEVICE=$reg_sp
 dnl AC_SUBST(CONFIG_HAVE_REGISTER_SOUND_SPECIAL_DEVICE)
 if test "$CONFIG_HAVE_REGISTER_SOUND_SPECIAL_DEVICE" = "1"; then
@@ -2660,20 +2362,7 @@
 dnl Check for old kill_fasync...
 AC_MSG_CHECKING(for old kill_fasync)
 oldkfasync="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$CFLAGS $KERNEL_INC $HACK_KERNEL_INC"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
-#ifdef __alpha__ /* it's required (at least for 2.2.18) */
-#include <asm/current.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/fs.h>
 ],[
  struct fasync_struct *fa;
@@ -2682,10 +2371,7 @@
 ],
     AC_MSG_RESULT(yes);oldkfasync="1",
     AC_MSG_RESULT(no);oldkfasync="0",
-    AC_MSG_RESULT(unknown);oldkfasync="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_OLD_KILL_FASYNC=$oldkfasync
 dnl AC_SUBST(CONFIG_OLD_KILL_FASYNC)
 if test "$CONFIG_OLD_KILL_FASYNC" = "1"; then
@@ -2695,27 +2381,14 @@
 dnl Check for dma_addr_t
 AC_MSG_CHECKING(for dma_addr_t)
 dma_addr_t="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$CFLAGS $KERNEL_INC $HACK_KERNEL_INC"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <asm/types.h>
 ],[
  dma_addr_t addr;
 ],
     AC_MSG_RESULT(yes);dma_addr_t="1",
     AC_MSG_RESULT(no);dma_addr_t="0",
-    AC_MSG_RESULT(unknown);dma_addr_t="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_DMA_ADDR_T=$dma_addr_t
 dnl AC_SUBST(CONFIG_HAVE_DMA_ADDR_T)
 if test "$CONFIG_HAVE_DMA_ADDR_T" = "1"; then
@@ -2725,17 +2398,7 @@
 dnl Check for MUTEX macros
 AC_MSG_CHECKING(for MUTEX macros)
 have_mutex_macros="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$CFLAGS $KERNEL_INC $HACK_KERNEL_INC"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <asm/semaphore.h>
 ],[
 #ifndef init_MUTEX
@@ -2745,10 +2408,7 @@
 ],
     AC_MSG_RESULT(yes);have_mutex_macros="1",
     AC_MSG_RESULT(no);have_mutex_macros="0",
-    AC_MSG_RESULT(unknown);have_mutex_macros="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_MUTEX_MACROS=$have_mutex_macros
 dnl AC_SUBST(CONFIG_HAVE_MUTEX_MACROS)
 if test "$CONFIG_HAVE_MUTEX_MACROS" = "1"; then
@@ -3005,17 +2665,7 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for class_simple)
   class_simple="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/device.h>
 ],[
   struct class_simple *(*foo)(struct module *, char *) = class_simple_create;
@@ -3023,10 +2673,7 @@
 ],
     AC_MSG_RESULT(yes);class_simple="1",
     AC_MSG_RESULT(no);class_simple="0",
-    AC_MSG_RESULT(unknown);class_simple="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_HAVE_CLASS_SIMPLE=$class_simple
   dnl AC_SUBST(CONFIG_SND_HAVE_CLASS_SIMPLE)
   if test "$CONFIG_SND_HAVE_CLASS_SIMPLE" = "1"; then
@@ -3038,17 +2685,7 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for old driver suspend/resume callbacks)
   old_driver_suspend="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/device.h>
 ],[
   static struct device_driver driver;
@@ -3056,10 +2693,7 @@
 ],
     AC_MSG_RESULT(yes);old_driver_suspend="1",
     AC_MSG_RESULT(no);old_driver_suspend="0",
-    AC_MSG_RESULT(unknown);old_driver_suspend="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_OLD_DRIVER_SUSPEND=$old_driver_suspend
   dnl AC_SUBST(CONFIG_SND_OLD_DRIVER_SUSPEND)
   if test "$CONFIG_SND_OLD_DRIVER_SUSPEND" = "1"; then
@@ -3071,27 +2705,14 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for removal of page-reservation for nopage/mmap)
   remove_page_reserve="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/mm.h>
 ],[
   void (*func)(struct vm_area_struct *, pte_t, unsigned long) = print_bad_pte;
 ],
     AC_MSG_RESULT(yes);remove_page_reserve="1",
     AC_MSG_RESULT(no);remove_page_reserve="0",
-    AC_MSG_RESULT(unknown);remove_page_reserve="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_REMOVE_PAGE_RESERVE=$remove_page_reserve
   dnl AC_SUBST(CONFIG_SND_REMOVE_PAGE_RESERVE)
   if test "$CONFIG_SND_REMOVE_PAGE_RESERVE" = "1"; then
@@ -3103,17 +2724,7 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for nested class_device)
   nested_class_device="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/device.h>
 ],[
   struct class_device dev;
@@ -3121,10 +2732,7 @@
 ],
     AC_MSG_RESULT(yes);nested_class_device="1",
     AC_MSG_RESULT(no);nested_class_device="0",
-    AC_MSG_RESULT(unknown);nested_class_device="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_NESTED_CLASS_DEVICE=$nested_class_device
   dnl AC_SUBST(CONFIG_SND_NESTED_CLASS_DEVICE)
   if test "$CONFIG_SND_NESTED_CLASS_DEVICE" = "1"; then
@@ -3136,17 +2744,7 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for new IRQ handler)
   new_irq_handler="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/interrupt.h>
 ],[
   irq_handler_t handle = 0;
@@ -3154,10 +2752,7 @@
 ],
     AC_MSG_RESULT(yes);new_irq_handler="1",
     AC_MSG_RESULT(no);new_irq_handler="0",
-    AC_MSG_RESULT(unknown);new_irq_handler="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_NEW_IRQ_HANDLER=$new_irq_handler
   dnl AC_SUBST(CONFIG_SND_NEW_IRQ_HANDLER)
   if test "$CONFIG_SND_NEW_IRQ_HANDLER" = "1"; then
@@ -3168,17 +2763,7 @@
 dnl Check for gfp_t...
 AC_MSG_CHECKING(for gfp_t)
 gfp_t="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/types.h>
 #include <linux/gfp.h>
 ],[
@@ -3186,10 +2771,7 @@
 ],
     AC_MSG_RESULT(yes);gfp_t="1",
     AC_MSG_RESULT(no);gfp_t="0",
-    AC_MSG_RESULT(unknown);gfp_t="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_GFP_T=$gfp_t
 dnl AC_SUBST(CONFIG_HAVE_GFP_T)
 if test "$CONFIG_HAVE_GFP_T" = "1"; then
@@ -3199,17 +2781,7 @@
 dnl Check for GFP_DMA32...
 AC_MSG_CHECKING(for GFP_DMA32)
 gfp_dma32="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/types.h>
 #include <linux/gfp.h>
 ],[
@@ -3217,10 +2789,7 @@
 ],
     AC_MSG_RESULT(yes);gfp_dma32="1",
     AC_MSG_RESULT(no);gfp_dma32="0",
-    AC_MSG_RESULT(unknown);gfp_dma32="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_GFP_DMA32=$gfp_dma32
 dnl AC_SUBST(CONFIG_HAVE_GFP_DMA32)
 if test "$CONFIG_HAVE_GFP_DMA32" = "1"; then
@@ -3230,17 +2799,7 @@
 dnl Check for page_to_pfn...
 AC_MSG_CHECKING(for page_to_pfn)
 page_to_pfn="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+KBUILD_TRY_COMPILE([
 #include <linux/types.h>
 #include <linux/mm.h>
 ],[
@@ -3250,10 +2809,7 @@
 ],
     AC_MSG_RESULT(yes);page_to_pfn="1",
     AC_MSG_RESULT(no);page_to_pfn="0",
-    AC_MSG_RESULT(unknown);page_to_pfn="0"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_HAVE_PAGE_TO_PFN=$page_to_pfn
 dnl AC_SUBST(CONFIG_HAVE_PAGE_TO_PFN)
 if test "$CONFIG_HAVE_PAGE_TO_PFN" = "1"; then
@@ -3266,20 +2822,7 @@
 else
 AC_MSG_CHECKING(for init_utsname)
   init_utsname="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-  if test "$CONFIG_HAVE_DEPRECATED_CONFIG_H" = "1"; then
-    CFLAGS="$CFLAGS -DCONFIG_HAVE_DEPRECATED_CONFIG_H=1"
-  fi
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/utsname.h>
 ],[
   int *x = (int *)init_utsname();
@@ -3287,10 +2830,8 @@
 ],
     AC_MSG_RESULT(yes);init_utsname="1",
     AC_MSG_RESULT(no);init_utsname="0",
-    AC_MSG_RESULT(unknown);init_utsname="0"
+    -Werror
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_HAVE_INIT_UTSNAME=$init_utsname
 fi
 dnl AC_SUBST(CONFIG_HAVE_INIT_UTSNAME)
@@ -3304,17 +2845,7 @@
 else
 AC_MSG_CHECKING(for is_power_of_2)
   is_power_of_2="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS -Werror"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/log2.h>
 ],[
   int a = is_power_of_2(2);
@@ -3322,10 +2853,8 @@
 ],
     AC_MSG_RESULT(yes);is_power_of_2="1",
     AC_MSG_RESULT(no);is_power_of_2="0",
-    AC_MSG_RESULT(unknown);is_power_of_2="0"
+    -Werror
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_HAVE_IS_POWER_OF_2=$is_power_of_2
 fi
 dnl AC_SUBST(CONFIG_HAVE_IS_POWER_OF_2)
@@ -3339,17 +2868,7 @@
 else
   AC_MSG_CHECKING(for __ffs)
   have_ffs="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/bitops.h>
 ],[
   unsigned long (*x)(unsigned long);
@@ -3357,10 +2876,7 @@
 ],
     AC_MSG_RESULT(yes);have_ffs="1",
     AC_MSG_RESULT(no);have_ffs="0",
-    AC_MSG_RESULT(unknown);have_ffs="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_HAVE_FFS=$have_ffs
 fi
 dnl AC_SUBST(CONFIG_HAVE_FFS)
@@ -3372,17 +2888,7 @@
 if test "$CONFIG_PNP_KERNEL" = "y"; then
   AC_MSG_CHECKING(for PnP suspend/resume)
   pnp_suspend="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/pnp.h>
 ],[
   struct pnp_driver pnp;
@@ -3390,10 +2896,7 @@
 ],
     AC_MSG_RESULT(yes);pnp_suspend="1",
     AC_MSG_RESULT(no);pnp_suspend="0",
-    AC_MSG_RESULT(unknown);pnp_suspend="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_HAVE_PNP_SUSPEND=$pnp_suspend
   dnl AC_SUBST(CONFIG_HAVE_PNP_SUSPEND)
   if test "$CONFIG_HAVE_PNP_SUSPEND" = "1"; then
@@ -3405,27 +2908,14 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for device_create_drvdata)
   device_create_drvdata="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/device.h>
 ],[
   struct device *(*foo)() = device_create_drvdata;
 ],
     AC_MSG_RESULT(yes);device_create_drvdata="1",
     AC_MSG_RESULT(no);device_create_drvdata="0",
-    AC_MSG_RESULT(unknown);device_create_drvdata="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_HAS_DEVICE_CREATE_DRVDATA=$device_create_drvdata
   if test "$CONFIG_SND_HAS_DEVICE_CREATE_DRVDATA" = "1"; then
     AC_DEFINE(CONFIG_SND_HAS_DEVICE_CREATE_DRVDATA)
@@ -3530,17 +3020,7 @@
 if test $kvernum -ge 206000; then
   AC_MSG_CHECKING(for new unlocked/compat_ioctl)
   new_ioctl="0"
-  ac_save_CFLAGS="$CFLAGS"
-  ac_save_CC="$CC"
-  CFLAGS="$KERNEL_CHECK_CFLAGS"
-  CC=$KCC
-  AC_TRY_COMPILE([
-#define __KERNEL__
-#ifdef CONFIG_HAVE_DEPRECATED_CONFIG_H
-#include <$INCLUDE_AUTOCONF_H>
-#else
-#include <linux/config.h>
-#endif
+  KBUILD_TRY_COMPILE([
 #include <linux/fs.h>
 ],[
   struct file_operations ops;
@@ -3549,10 +3029,7 @@
 ],
     AC_MSG_RESULT(yes);new_ioctl="1",
     AC_MSG_RESULT(no);new_ioctl="0",
-    AC_MSG_RESULT(unknown);new_ioctl="0"
 )
-  CFLAGS=$ac_save_CFLAGS
-  CC=$ac_save_CC
   CONFIG_SND_HAVE_NEW_IOCTL=$new_ioctl
   dnl AC_SUBST(CONFIG_SND_HAVE_NEW_IOCTL)
   if test "$CONFIG_SND_HAVE_NEW_IOCTL" = "1"; then
@@ -3564,20 +3041,13 @@
 dnl compiler support for _Bool
 AC_MSG_CHECKING(for builtin _Bool support)
 builtin_bool="0"
-ac_save_CFLAGS="$CFLAGS"
-ac_save_CC="$CC"
-CFLAGS="$KERNEL_CHECK_CFLAGS"
-CC=$KCC
-AC_TRY_COMPILE([],[
+KBUILD_TRY_COMPILE([],[
   _Bool b;
   b = (_Bool)0;
 ],
   AC_MSG_RESULT(yes);builtin_bool="1",
   AC_MSG_RESULT(no);builtin_bool="0",
-  AC_MSG_RESULT(unknown);builtin_bool="1"
 )
-CFLAGS=$ac_save_CFLAGS
-CC=$ac_save_CC
 CONFIG_SND_HAS_BUILTIN_BOOL=$builtin_bool
 if test "$CONFIG_SND_HAS_BUILTIN_BOOL" = "1"; then
   AC_DEFINE(CONFIG_SND_HAS_BUILTIN_BOOL)
